### DISCUSSION TITLE
Can you add a DS18B20 temperature probe to any Sonoff device with an accessible GPIO pin?

### DISCUSSION THREAD

**Discussion by wayner9** [arendst/Tasmota#19968] [General]
I have read about adding temperature sensors to various devices.  Can you add these probes to any GPIO port?  Then you just configure the port and it will be reading the temperature?  Are there only specific GPIO ports that work for this?

**Comment by barbudor** [comment]
It depends...
On Esp8266 some pins have some constraint as they are used to configure the processor boot mode. So their are pull up on GPIO0 and GPIO2 and whatever you connector to such pins should not drive them low at boot. Opposite for GPIO15 which as a pull down and should not be high on boot
Also GPIO1 (serial tx) should not be tighed to low at boot
On ESP32 you may have more flexibility

But you may also be limited by the gpios that a given device leave free and available. 

Also NEVER try to connect sensors to a power monitoring device. This devices are NOT isolated from the mains line and there is a high risk of electrocution if a sensor is connected outside of the box. 

So the best would be to start by
- do you already have devices, if so which ones? 
- if not and are looking for advices what are you looking for apart from attaching sensors? 

If your only interrest is about sensor, better look into dev boards like NodeMCU or more recent boards based on ESP32-C3

**Reply by wayner9** [reply]
One device is the Sonoff Basic R2.  When I flashed one of these last week I soldered a header onto the circuit board into five holes - Vcc, Tx, RX, Gnd and GPIO14.

The other device that I would like to attach a thermometer to is a Sonoff POWR3.  It has two pads labelled GPIO015 and GPIO016.  It may have another GPIO pin accessible on a header, as it does have a header which makes it very easy to flash as you don't need to solder.

**Reply by barbudor** [reply]
Basic R2 : Fine to use GPIO14 for DS18x20

POWR3: As I said do not connect anything to power monitoring devices. These are not isolated.

**Reply by wayner9** [reply]
Ok thanks, I misunderstood.  I thought you meant that don't connect to the power monitoring pins.  That would especially be a risk with this switch since it can handle 5500W so that is a lot of power that could be very dangerous.

**Comment by Th0maz** [comment]
> One device is the Sonoff Basic R2.

I've got a handful of those devices running for a couple of years now with a BME280 sensor attached.
I chose them because they were cheap, came with an integrated 230V power supply and had some spare GPIO pins.
3D-printed a matching enclosure with cable outlets to put them in every room.

I know these devices have a very poor power supply, but since I don't use the relay, they seem to get along.

**Reply by wayner9** [reply]
What do you mean by a poor power supply?  You mean the AC to DC converter within the unit?  I wonder if it makes any difference between being on 120V vs 240V mains as I am in Canada where we have 120V.

**Reply by sfromis** [reply]
Yes, it can barely provide itself with enough 3.3V DC power, and is not rare that people have trouble. IDK if the lower AC voltage ones are equally bad. I'd not recommend Sonoff Basic, especially the R2, as reasonable devices to add stuff to. Some are known to die even without such extra loads; I've have had two of mine stop working.

**Reply by wayner9** [reply]
Thanks - what devices would you suggest?  Or should you even get a device that has a temperature probe functionality built in, like the THR316 or THR316D since these are already pretty cheap.  Even if you don't need the switch capability.  

For example. the THR316D is C$17.99 on Amazon Canada with an $8 off coupon.  And it has a screen.  I was thinking of configuring an RPi to use a temp monitor, but these are so cheap it may be much easier to just use that.  I have some RJ9 connectors so I just made a pigtail of CAT5e to RJ9, and then just connect wires from the DS18B20 to the CAT5e wires.  

Or if you have an old phone handset with an RJ9 cord on the handset you can cut that cord and use it, as long as it has four wires.

**Reply by sfromis** [reply]
My sensor devices are built on dev boards, with USB power and sometimes a cheap electronics box around it. Also provides good opportunities for adding sensors for CO2, PM, light or movement, and maybe a small display too, or neopixels to provide a simple indicator.

**Reply by wayner9** [reply]
Like these:  https://www.amazon.ca/ESP-WROOM-32-Development-Bluetooth-Microcontroller-Arduino/dp/B0C9VSYS4N/

Or what else would you recommend?  I will get a couple of these to play around.  But getting the THR devices for $25 with a display isn't a bad way to go for a smart thermometer and you get a switch and dry contact relay for free!

**Reply by sfromis** [reply]
I'm not doing recommendations, as things depends on circumstances, but what you linked is one of many dev boards usable for such. And I'm normally buying at much lower prices from China, where similar boards are around US$4 and up.

**Reply by wayner9** [reply]
Ok thanks.  Those are roughly C$11 each which is US$7 so paying a bit more for next day delivery makes sense.

**Comment by lalo-uy** [comment]
The Badic have very little power to spare, a DS18b20 only needs  cuple of
mA while active, plus the pull up .  Use 3k3 or lower.
Them use to get stuck from time to time any way.

**Comment by lalo-uy** [comment]
I have use Sonoff Dual with good results for beer fermenters. Have 2 outpus
for heat and cold, and 18B20 where stable on 3m cords. The case has the 3mm
hole for the wire.
